src/test/kotlin/com/fwdekker/randomness/SettingsTest.kt

```diff
@@ -57,3 +77,212 @@ object SettingsTest : FunSpec({
+
+                shouldNotThrowAny { persistent.upgrade(stored) }
+            }
+        }
+
+        context("generic behaviour") {
+            test("bumps the version number if the target version is higher than the stored version") {
+                val stored = """<component><option name="version" value="3.1.0"/></component>""".parseXml()
+                stored.getPropertyValue("version") shouldBe "3.1.0"
+
+                val patched = persistent.upgrade(stored, Version.parse("3.2.0"))
+                patched.getPropertyValue("version") shouldBe "3.2.0"
+            }
+
+            test("does not bump the version number if the target version is lower than the stored version") {
+                val stored = """<component><option name="version" value="3.1.0"/></component>""".parseXml()
+                stored.getPropertyValue("version") shouldBe "3.1.0"
+
+                val patched = persistent.upgrade(stored, Version.parse("3.0.0"))
+                patched.getPropertyValue("version") shouldBe "3.1.0"
+            }
+
+            test("applies multiple upgrades in sequence if needed") {
+                val stored = getTestConfig("/settings-upgrades/v3.1.0-v3.3.5.xml").parseXml()
+                stored.getSchemes().single().run {
+                    getPropertyValue("type") shouldBe "1"
+                    getProperty("version") shouldBe null
+                }
+                stored.getDecorators() shouldNot beEmpty()
+                stored.getDecorators().forEach { it.getProperty("generator") shouldNotBe null }
+
+                val patched = persistent.upgrade(stored, Version.parse("3.3.5"))
+                patched.getSchemes().single().run {
+                    getProperty("type") shouldBe null
+                    getPropertyValue("version") shouldBe "1"
+                }
+                patched.getDecorators() shouldNot beEmpty()
+                patched.getDecorators().forEach { it.getProperty("generator") shouldBe null }
+            }
+
+            test("upgrades only up to the specified version") {
+                val stored = getTestConfig("/settings-upgrades/v3.1.0-v3.3.5.xml").parseXml()
+                stored.getSchemes().single().run {
+                    getPropertyValue("type") shouldBe "1"
+                    getProperty("version") shouldBe null
+                }
+                stored.getDecorators() shouldNot beEmpty()
+                stored.getDecorators().forEach { it.getProperty("generator") shouldNotBe null }
+
+                val patched = persistent.upgrade(stored, Version.parse("3.2.0"))
+                patched.getSchemes().single().run {
+                    getProperty("type") shouldBe null
+                    getPropertyValue("version") shouldBe "1"
+                }
+                stored.getDecorators() shouldNot beEmpty()
+                stored.getDecorators().forEach { it.getProperty("generator") shouldNotBe null }
+            }
```

Comment:
Add test: Does not bump version if no upgrades are available. E.g. does not bump to v3.1.1 since there is no format version v3.1.1.